src/VisualStudio/Core/Impl/ProjectSystem/CPS/CPSProject_IWorkspaceProjectContext.cs

```diff
@@ -58,8 +58,6 @@ public bool LastDesignTimeBuildSucceeded
 
         public CPSProject(VisualStudioProject visualStudioProject, VisualStudioWorkspaceImpl visualStudioWorkspace, IProjectCodeModelFactory projectCodeModelFactory, Guid projectGuid, string binOutputPath)
         {
-            ThreadHelper.ThrowIfNotOnUIThread();
```

Comment:
Was this an incorrect assert?

Reply:
The code in this constructor doesn't have UI affinity, but calls SubscribeExternalErrorDiagnosticUpdateSourceToSolutionBuildEvents that already would throw. If we fix up that subscription to no longer require the UI thread then this would be an incorrect assert. In this case I'm erring on there's no reason to have a function capture all of the transitive assert requirements.

Reply:
> there's no reason to have a function capture all of the transitive assert requirements

I would prefer if it did. Looking at a function I want to know whether it should be run on UI thread. If it doesn't declare that by explicitly calling `ThreadHelper.ThrowIfNotOnUIThread` or similar API then I have no idea without checking all functions it calls and their transitive closure, which is not feasible.